tests/unit/s2n_fingerprint_ja4_test.c

```diff
@@ -0,0 +1,1258 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "api/unstable/fingerprint.h"
+#include "crypto/s2n_fips.h"
+#include "s2n_test.h"
+#include "testlib/s2n_sslv2_client_hello.h"
+#include "testlib/s2n_testlib.h"
+#include "tls/s2n_tls.h"
+
+#define S2N_TEST_OUTPUT_SIZE 200
+
+#define S2N_TEST_CLIENT_HELLO_VERSION \
+    0x00, 0x00
+
+/* clang-format off */
+#define S2N_TEST_CLIENT_HELLO_AFTER_VERSION \
+    /* random */ \
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, \
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, \
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, \
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, \
+    /* session id */ \
+    0x00
+/* clang-format on */
+
+/* clang-format off */
+#define S2N_TEST_CLIENT_HELLO_CIPHERS \
+    /* cipher suites size */ \
+    0x00, 0x02, \
+    /* cipher suites */ \
+    0x00, 0x00
+/* clang-format on */
+
+#define S2N_TEST_CLIENT_HELLO_AFTER_CIPHERS \
+    /* legacy compression methods */        \
+    0x01, 0x00
+
+#define S2N_TEST_CLIENT_HELLO_EMPTY_EXTENSIONS \
+    0x00, 0x00
+#define S2N_TEST_CLIENT_HELLO_EMPTY_EXTENSION \
+    0x00, 0x00
+
+/* This macro currently assumes that the message size is only one byte (<=255). */
+#define S2N_INIT_CLIENT_HELLO(name, ...)                 \
+    uint8_t _##name##_message[] = { __VA_ARGS__ };       \
+    EXPECT_TRUE(sizeof(_##name##_message) <= UINT8_MAX); \
+    uint8_t name[] = {                                   \
+        TLS_CLIENT_HELLO,                                \
+        0x00, 0x00, sizeof(_##name##_message),           \
+        __VA_ARGS__                                      \
+    }
+
+/* clang-format off */
+#define S2N_INIT_CLIENT_HELLO_WITH_EXTENSION(name, extension) \
+    S2N_INIT_CLIENT_HELLO(name, \
+        S2N_TEST_CLIENT_HELLO_VERSION, \
+        S2N_TEST_CLIENT_HELLO_AFTER_VERSION, \
+        S2N_TEST_CLIENT_HELLO_CIPHERS, \
+        S2N_TEST_CLIENT_HELLO_AFTER_CIPHERS, \
+        /* extensions size */ \
+        0x00, 0x04, \
+        /* extension */ \
+        0x00, extension, S2N_TEST_CLIENT_HELLO_EMPTY_EXTENSION \
+    )
+/* clang-format on */
+
+enum {
+    S2N_JA4_A_PROTOCOL = 0,
+    S2N_JA4_A_VERSION_1,
```

Comment:
```suggestion
    S2N_JA4_A_VERSION_FIRST_BYTE,
    S2N_JA4_A_VERSION_SECOND_BYTE,
```
What about something like this instead? Presumably these are referring to the location of each value?